mm2src/coins/utxo/utxo_tests.rs

```diff
@@ -4265,3 +4265,90 @@ fn test_utxo_validate_valid_and_invalid_pubkey() {
     assert!(coin.validate_other_pubkey(&[1u8; 20]).is_err());
     assert!(coin.validate_other_pubkey(&[1u8; 8]).is_err());
 }
+
+#[test]
+fn test_block_header_utxo_loop() {
+    use crate::utxo::utxo_builder::{block_header_utxo_loop, BlockHeaderUtxoLoopExtraArgs};
+    use futures::future::{Either, FutureExt};
+
+    static mut CURRENT_BLOCK_COUNT: u64 = 150;
+
+    ElectrumClient::get_block_count
+        .mock_safe(move |_| MockResult::Return(Box::new(futures01::future::ok(unsafe { CURRENT_BLOCK_COUNT }))));
+
+    BlockHeaderUtxoLoopExtraArgs::default.mock_safe(move || {
+        MockResult::Return(BlockHeaderUtxoLoopExtraArgs {
+            chunk_size: 100,
+            error_sleep: 5.,
+            success_sleep: 1.,
+        })
+    });
+
+    let block_header_storage = BlockHeaderStorageForTests::new(TEST_COIN_NAME.to_string());
+    BlockHeaderStorage::new_from_ctx.mock_safe(move |_, _| {
+        MockResult::Return(Ok(BlockHeaderStorage {
+            inner: Box::new(block_header_storage.clone()),
+        }))
+    });
+
+    let ctx = MmCtxBuilder::new().into_mm_arc();
+    let priv_key_policy = PrivKeyBuildPolicy::IguanaPrivKey(&[1u8; 32]);
+
+    let mut args = ElectrumBuilderArgs::default();
+    args.collect_metrics = false;
+    args.spawn_ping = false;
+
+    let servers: Vec<_> = [
+        "electrumx1.cointest.com:50001",
+        "electrumx2.cointest.com:50001",
+        "electrumx3.cointest.com:50001",
+    ]
+    .iter()
+    .map(|server| json!({ "url": server,"disable_cert_verification":true }))
+    .collect();
+    let servers = servers.into_iter().map(|s| json::from_value(s).unwrap()).collect();
+    let conf = json!({"coin":"WHIVE","asset":"WHIVE","rpcport":8923,"enable_spv_proof": false});
+    let req = json!({ "method": "electrum", "servers": servers });
+
+    let params = UtxoActivationParams::from_legacy_req(&req).unwrap();
+    let builder = UtxoArcBuilder::new(&ctx, "WHIVE", &conf, &params, priv_key_policy, UtxoStandardCoin::from);
+    let client = block_on(builder.electrum_client(AbortableQueue::default(), args, servers)).unwrap();
+    let arc: UtxoArc = block_on(builder.build_utxo_fields()).unwrap().into();
+
+    let (sync_status_notifier, _) = channel::<UtxoSyncStatus>(1);
+    let loop_handle = UtxoSyncStatusLoopHandle::new(sync_status_notifier);
+
+    let loop_fut = async move {
+        unsafe {
+            block_header_utxo_loop(
+                arc.downgrade(),
+                UtxoStandardCoin::from,
+                loop_handle,
+                CURRENT_BLOCK_COUNT,
+            )
+            .await
+        }
+    };
+
+    let test_fut = async move {
+        Timer::sleep(3.).await;
+        let get_headers_count = client.block_headers_storage().get_last_block_height().await.unwrap();
+        assert_eq!(150, get_headers_count);
```

Comment:
This test can be simplified by setting the block heights to lower values.
For example, 
1. Set `chunks_size = 4`, `CURRENT_BLOCK_COUNT = 13` initially (so here we check that the loop fetches the first 13 blocks by [1, 4], [5, 8], [9, 12], [13] steps).
2. Set `CURRENT_BLOCK_COUNT = 17` (here we check that the loop requests from 13 to 17 blocks by [14, 17])
3. Set `CURRENT_BLOCK_COUNT = 18` (in this case we check that the loop requests the only 18th block)
4. Set `CURRENT_BLOCK_COUNT = 25` (here we check the following block:
```rust
// More than `chunk_size` blocks could have appeared since the last `get_block_count` RPC.
// So reset `to_block_height` if only `from_block_height + chunk_size > actual_block_count`.
if to_block_height > block_count {
  to_block_height = block_count;
}
```
The steps expected: [19, 22], [23, 25], so not in one step).

This is important - the test needs to validate that we request only those blocks **that we need** by the steps we require.

It could be done via making `ElectrumClient::blockchain_block_headers` mockable
```rust
let expected_steps = Arc::new(Mutex::new(vec![]));
    let expected_steps_to_move = expected_steps.clone();
    ElectrumClient::retrieve_headers.mock_safe(move |this, from, count| {
        let (expected_from, expected_count) = expected_steps_to_move.lock().unwrap().remove(0);
        assert_eq!(from, expected_from);
        assert_eq!(count, expected_count);
        // Continue `ElectrumClient::retrieve_headers` executing.
        MockResult::Continue((this, from, count))
    });

    // In the test async block
    let test_fut = async move {
        // The pairs where first is `from` and second is `count`.
        *expected_steps.lock().unwrap() = vec![(1, 4), (5, 4), (9, 4), (13, 1)];
        unsafe { CURRENT_BLOCK_COUNT = 13 };

        Timer::sleep(3.).await;
        let get_headers_count = client.block_headers_storage().get_last_block_height().await.unwrap();
        assert_eq!(get_headers_count, 13);
        // Check that we got to all steps.
        assert!(expected_steps.lock().unwrap().is_empty());
    };
```

Hope it's clear :)

Reply:
yep!